File: javascript/core/src/verbs/fetch.ts
/*!
 * Copyright (c) Microsoft. All rights reserved.
 * Licensed under the MIT license. See LICENSE file in the project.
 */
import type { TableContainer } from '@datashaper/arquero'
import { container } from '@datashaper/arquero'
import type { InputStep } from '@datashaper/dataflow'
import { inputNodeFactory } from '@datashaper/dataflow'
import type { FetchArgs } from '@datashaper/schema'
import { loadCSV, loadJSON } from 'arquero'

export const fetchStep: InputStep<TableContainer, FetchArgs> = async (
	{ url, delimiter, autoMax },
	id,
) => {
	const table = url.toLowerCase().endsWith('.json')
		? loadJSON(url, {
				autoType: autoMax === undefined || autoMax <= 0 ? false : true,
		  })
		: loadCSV(url, {
				delimiter,
				autoMax: autoMax !== undefined ? autoMax : 0,
				autoType: autoMax === undefined || autoMax <= 0 ? false : true,
		  })
	return container(id, await table)
}

export const fetch = inputNodeFactory(fetchStep)

File: python/datashaper/datashaper/engine/verbs/fetch.py
#
# Copyright (c) Microsoft. All rights reserved.
# Licensed under the MIT license. See LICENSE file in the project.
#

from urllib.parse import urlparse

import pandas as pd

from datashaper.table_store import TableContainer


__reader_mapping = {
    "csv": lambda **kwargs: pd.read_csv(
        kwargs["url"], sep=kwargs["delimiter"], na_values=["undefined"]
    ),
    "json": lambda **kwargs: pd.read_json(kwargs["url"]),
}


def fetch(url: str, delimiter: str = ",", **kwargs):
    file_type = urlparse(url).path.split(".")[-1]
    output = __reader_mapping[file_type](**{"url": url, "delimiter": delimiter})
    return TableContainer(table=output)
